plugins/scale/scale-title-overlay.cpp

```diff
@@ -0,0 +1,326 @@
+#include "scale-title-overlay.hpp"
+
+#include <wayfire/opengl.hpp>
+#include <wayfire/plugins/common/cairo-util.hpp>
+#include <wayfire/plugins/common/simple-texture.hpp>
+
+class view_title_overlay : public wf::custom_data_t
+{
+  protected:
+    wf::cairo_text_t overlay;
+    wf::cairo_text_t::params par;
+    wayfire_view view;
+    /* the transformer we are attached to */
+    wf::scale_transformer& tr;
+    /* if text potentially overflows the available space */
+    bool overflow = false;
+
+    /**
+     * Get the transformed bounding box of the view, including the
+     * current transform
+     */
+    wlr_box get_transformed_bounding_box()
+    {
+        wlr_box box = view->get_wm_geometry();
+        box = tr.trasform_box_without_padding(box);
+        return box;
+    }
+
+    void render_overlay()
+    {
+        wlr_box box = get_transformed_bounding_box();
+        par.max_size = {box.width, box.height};
+        auto res = overlay.render_text(view->get_title().c_str(), par);
+        overflow = res.width > overlay.tex.width;
+        view->damage();
+    }
+
+  public:
+    enum class position
+    {
+        TOP,
+        CENTER,
+        BOTTOM,
+    };
+
+    position pos = position::CENTER;
+
+    GLuint get_texture(float output_scale, wf::geometry_t& geometry)
+    {
+        wlr_box box = get_transformed_bounding_box();
+        int w = overlay.tex.width;
+        int h = overlay.tex.height;
+
+        /**
+         * regenerate the overlay texture in the following cases:
+         * 1. Output's scale changed
+         * 2. The overlay does not fit anymore
+         * 3. The overlay previously did not fit, but there is more space now
+         * TODO: check if this wastes too high CPU power when views are being
+         * animated and maybe redraw less frequently or in a way that will not
+         * hold up the rendering loop!
+         */
+        if ((output_scale != par.output_scale) || (w > box.width * output_scale) ||
+            (overflow && (w < std::floor(box.width * output_scale))))
+        {
+            par.output_scale = output_scale;
+            par.max_size     = {box.width, box.height};
+            auto res = overlay.render_text(view->get_title().c_str(), par);
+            view->damage();
+            w = overlay.tex.width;
+            h = overlay.tex.height;
+            overflow = res.width > w;
+        }
+
+        int y = 0;
+        switch (pos)
+        {
+          /* TODO: Add some padding! */
```

Comment:
add padding where?

Reply:
Between the view and the title label (if it is above or below; in these cases, it is exactly touching it). I'm not sure if this is actually necessary, it does not look as bad as I expected.

Reply:
Alright. We can leave it for later if someone complains :D